extension/src/commands.ts

```diff
@@ -32,12 +32,63 @@ export function registerCommands(context: vscode.ExtensionContext){
 		}
 		catch (err) {
 			if (err){
-				let msg = localize("autolispext.commands.addFoldingRegion", "Failed to insert snippet");
+				let msg = AutoLispExt.localize("autolispext.commands.addFoldingRegion", "Failed to insert snippet");
 				showErrorMessage(msg, err);
 			}
 		}
 	}));
 
 
+	// Associated with the right click "Generate Documentation" menu item
+	context.subscriptions.push(vscode.commands.registerCommand('autolisp.generateDocumentation', async () => {
+		try {
+			const pos = vscode.window.activeTextEditor.selection.start;
+			const doc = AutoLispExt.Documents.getDocument(vscode.window.activeTextEditor.document);
+			const exp = doc.atomsForest.find(p => p.contains(pos));
+
+			// Locate all the Defun & Defun-Q statements or an empty array if no actionable scope was found
+			let defs = exp?.body?.findChildren(shared.SearchPatterns.DEFINES, true).filter(p => p.contains(pos)) ?? [];
+			if (defs.length === 0) {
+				return;
+			}
+			if (defs.length > 1) {
+				// if this was performed within a nested defun, then the user has to select which defun to document
+				const quickPicks = defs.map(d => d.getNthKeyAtom(1).symbol);
+				let outResult = '';
+				await vscode.window.showQuickPick(quickPicks).then(response => { outResult = response; });
+				defs = defs.filter(d => d.getNthKeyAtom(1).symbol === outResult);
+			}			
+			// Now we know what defun we are documenting, so we extract arguments; if applicable
+			const def = defs[0];
+			let args = def.atoms.find(p => p instanceof LispContainer).body?.atoms.filter(a => !a.isComment()) ?? [];
+			const dividerIndex = args.findIndex(a => a.symbol === '/');
+			if (dividerIndex === -1 && args.length > 2) {
+				args = args.slice(1, args.length - 1);
+			} else if (dividerIndex === -1) {
+				args = [];
+			} else {
+				args = args.slice(1, dividerIndex);
+			}
+
+			// generate a dynamic snippet multi-line comment to represent the defun and its arguments
+			const lf = vscode.window.activeTextEditor.document.eol === vscode.EndOfLine.LF ? '\n' : '\r\n';
+			const defPos = def.getRange().start;
+			let count = 1;
+			let insert = `${lf};|${lf}  $` + `{${count++}:description}${lf}`;
+			args.forEach(a => {
+				insert += `  @Param ${a.symbol} $` + `{${count++}:?}${lf}`;
+			});
+			insert += '  @Returns ${' + `${count}:type?}${lf}|;${lf}`;
+			const snip = new vscode.SnippetString(insert);
+			vscode.window.activeTextEditor.insertSnippet(snip, defPos);
+		}
+		catch (err) {
+			if (err) {
+				let msg = AutoLispExt.localize("autolispext.help.commands.generateDocumentation", "A valid defun name could not be located");
```

Comment:
@hualin-whl 
shall we add an item for this string in [the_owner_file].i18n.json by hand? 
e.g. if this code is kept here in this file, we then need change i18n\enu\out\commands.i18n.json. right?

Reply:
In the Objective section, I suggested we get square with i18 in a final (or 3rd) PR just because it will muddy this up quite a bit and make it harder to sift through what needs reviewing. I will do it all up front if needed, just trying to be kind 😃

Reply:
I added them, but of course that and your other comments to break stuff up has now blown up the number of changed files. I was trying to avoid that and do it incrementally, but its there now.